python/sparkdl/estimators/keras_image_file_estimator.py

```diff
@@ -0,0 +1,262 @@
+            model_filename = bytes_to_h5file(model_bytes)
+            transformers.append({
+                'paramMap': paramMap, 
+                'transformer': KerasImageFileTransformer(modelFile=model_filename)})
+
+        return transformers
+
+    def _fitInParallel(self, dataset, paramMaps):
+        """
+        Fits len(paramMaps) models in parallel, one in each Spark task.
+        :param paramMaps: non-empty list or tuple of ParamMaps (dict values)
+        :return: list of fitted models, matching the order of paramMaps
+        """
+        #sc = dataset.sql_ctx._sc
+        print('DataSet in _fitInParallel')
+        dataset.printSchema()
+        print('parameters (paramMaps)', paramMaps)
+
+        sc = JVMAPI._curr_sc()
+        paramMapsRDD = sc.parallelize(paramMaps, numSlices=len(paramMaps))
+
+        localFeatures, localLabels = self._getNumpyFeaturesAndLabels(dataset)
+        localFeaturesBc = sc.broadcast(localFeatures)
+        localLabelsBc = None if localLabels is None else sc.broadcast(localLabels)
+        modelBytes = self._loadModelAsBytes()
+        modelBytesBc = sc.broadcast(modelBytes)
+
+        print('Broadcast parameters created')
+        
+        kerasModelsRDD = paramMapsRDD.map(
+            lambda paramMap: (paramMap,
+                              self._localFit(localFeaturesBc, localLabelsBc, modelBytesBc, paramMap)))
+        return self._collectModels(kerasModelsRDD)
+
+    def _loadModelAsBytes(self):
+        """
+        (usable on driver only)
+        Load the Keras model file as a byte string.
+        :return: str containing the model data
+        """
+        with open(self.getModelFile(), mode='rb') as file:
+            fileContent = file.read()
+        return fileContent
+
+    def _fit(self, dataset):
+        err_msgs = ["This function should not have been called",
+                    "Please contact library mantainers to file a bug"]
+        raise NotImplementedError('\n'.join(err_msgs))
+
+    def _validateFitParams(self, params):
+        """ Check if an input parameter set is valid """
+        if isinstance(params, (list, tuple, dict)): 
+            assert self.getInputCol() not in params, \
+                "params {} cannot contain input column {}".format(params, self.getInputCol())
+        
+    def fit(self, dataset, params=None):
+        """
+        Fits a model to the input dataset with optional parameters.
+        :param dataset: input dataset, which is an instance of :py:class:`pyspark.sql.DataFrame`.
+                        The column `inputCol` should be of type
+                        `sparkdl.image.imageIO.imgSchema`.
+        :param params: An optional param map that overrides embedded params. If a list/tuple of
+                       param maps is given, this calls fit on each param map and returns a list of
+                       models.
+        :returns: fitted model(s).  If params includes a list of param maps, the order of these
```

Comment:
This API is not great because it returns a list of all the evaluated models. If you look at the other meta estimators in MLlib, they return the best one. Since the model evaluation is currently embedded in Keras (and outside the reach of Spark for now), this compromise is acceptable, but we will need to figure out how to make it return the best model eventually. Can you put a comment saying that this return value is subject to change in the future?

cc @jkbradley who may have some thoughts on this point.

Reply:
Discussed offline with @jkbradley and @thunterdb.